GVFS/GVFS.Service/RepoAutoMounter.cs

```diff
@@ -0,0 +1,150 @@
+﻿using System;
+using GVFS.Common;
+using GVFS.Common.FileSystem;
+using GVFS.Common.NamedPipes;
+using GVFS.Common.Tracing;
+using GVFS.Service.Handlers;
+
+namespace GVFS.Service
+{
+    public class RepoAutoMounter : IDisposable
+    {
+        private readonly ITracer tracer;
+        private readonly RepoRegistry repoRegistry;
+        private readonly int sessionId;
+        private readonly GVFSMountProcessManager mountProcessManager;
+        private readonly string userSid;
+
+        private IVolumeStateWatcher volumeWatcher;
+
+        public RepoAutoMounter(ITracer tracer, RepoRegistry repoRegistry, int sessionId)
+        {
+            this.tracer = tracer ?? throw new ArgumentNullException(nameof(tracer));
+            this.repoRegistry = repoRegistry ?? throw new ArgumentNullException(nameof(repoRegistry));
+            this.sessionId = sessionId;
+
+            // Create a mount process factory for this session/user
+            this.mountProcessManager = new GVFSMountProcessManager(this.tracer, sessionId);
+            this.userSid = this.mountProcessManager.CurrentUser.Identity.User?.Value;

```

Comment:
Have you ever seen `this.mountProcessManager.CurrentUser.Identity.User` be null?

Reply:
I have not seen it to be null, but some helpful(?) tooling indicated this property _could possibly_ be null.

https://docs.microsoft.com/en-us/dotnet/api/system.security.principal.windowsidentity.user?view=netframework-4.7.2#remarks

It appears the `WindowsIdentity.User` property is `null` only if the identity is the 'anonymous' user. Sessions could be created for anonymous users (file share connections open to 'everyone' perhaps?).. not sure if the fact we only act for `SessionChangeReason.SessionLogon` in `OnSessionChanged` means we don't see anonymous user sessions.

I can remove the `?.` if you'd like.

Reply:
What would this do if User is null? What does it mean for us to store a userSid of 0? How will that behave when we try to mount later?

The issue isn't so much the presence of the `?` as it is that this will likely lead to cascading failures down the line. If you're going to handle null users here, then you also need to do something other than carry on as if it succeeded.

Reply:
If user is null, then we'd cascade the nulls and have a `(string)null` SID. The call later when mounting to get the list of registered, active repositories for that user's SID would return no repositories (it filters the collection by `string.Compare(repo.OwnerSID, ownerSID, ..)`.

I guess this is a silent event that we might want to know about (that we have a null/anonymous user). I'll remove the `?.` to revert to the existing behaviour?